libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m

```diff
@@ -95,13 +96,35 @@ - (void)completeWithSuccess:(SFOAuthCredentials *)credentials {
 
 - (void)completeWithError:(NSError *)error {
     [self log:SFLogLevelError format:@"%@ Refresh failed with error: %@", NSStringFromSelector(_cmd), error];
-    if (self.errorBlock) {
+
+    if ([error.domain isEqualToString:kSFOAuthErrorDomain] && error.code == kSFOAuthErrorInvalidGrant) {
+        [self log:SFLogLevelInfo format:@"%@ Invalid grant error received, triggering logout.", NSStringFromSelector(_cmd)];
+        // make sure we call logoutUser on main thread
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self createAndStoreLogoutEvent:error];
+            [self logoutUser];
+        });
+    }
+    else if (self.errorBlock) {
         dispatch_async(dispatch_get_main_queue(), ^{
             self.errorBlock(error);
         });
     }
 }
 
+- (void)logoutUser {
+    SFUserAccountIdentity *userIdentity = [[SFUserAccountIdentity alloc] initWithUserId:self.coordinator.credentials.userId orgId:self.coordinator.credentials.organizationId];
```

Comment:
We have multiple consumers of this class though. Wouldn't it be nice for the session refresher to handle the results of that, instead of each consumer handling it, especially in the logout scenario?